data_safe_haven/pulumi/components/shm_firewall.py

```diff
@@ -0,0 +1,1128 @@
+"""Pulumi component for SHM traffic routing"""
+# Third party imports
+from pulumi import ComponentResource, Input, Output, ResourceOptions
+from pulumi_azure_native import network
+
+# Local imports
+from data_safe_haven.external.interface import AzureIPv4Range
+
+
+class SHMFirewallProps:
+    """Properties for SHMFirewallComponent"""
+
+    def __init__(
+        self,
+        location: Input[str],
+        resource_group_name: Input[str],
+        route_table_name: Input[str],
+        subnet_firewall_name: Input[str],
+        subnet_firewall_virtual_network_name: Input[str],
+        subnet_identity_iprange: Input[AzureIPv4Range],
+        subnet_update_servers_iprange: Input[AzureIPv4Range],
+    ):
+        self.location = location
+        self.resource_group_name = resource_group_name
+        self.route_table_name = route_table_name
+        self.subnet_firewall_name = subnet_firewall_name
+        self.subnet_firewall_virtual_network_name = subnet_firewall_virtual_network_name
+        self.subnet_identity_iprange = Output.from_input(subnet_identity_iprange).apply(
+            lambda ip_range: str(ip_range)
+        )
+        self.subnet_update_servers_iprange = Output.from_input(
+            subnet_update_servers_iprange
+        ).apply(lambda ip_range: str(ip_range))
+
+
+class SHMFirewallComponent(ComponentResource):
+    """Deploy SHM routing with Pulumi"""
+
+    def __init__(
+        self,
+        name: str,
+        stack_name: str,
+        shm_name: str,
+        props: SHMFirewallProps,
+        opts: ResourceOptions = None,
+    ):
+        super().__init__("dsh:shm:SHMFirewallComponent", name, {}, opts)
+        child_opts = ResourceOptions(parent=self)
+
+        # Important IP addresses
+        external_dns_resolver = "168.63.129.16"  # https://docs.microsoft.com/en-us/azure/virtual-network/what-is-ip-address-168-63-129-16
+        ntp_ip_addresses = [
+            "216.239.35.0",
+            "216.239.35.4",
+            "216.239.35.8",
+            "216.239.35.12",
+        ]
+        ntp_fqdns = [
+            "time.google.com",
+            "time1.google.com",
+            "time2.google.com",
+            "time3.google.com",
+            "time4.google.com",
+        ]
+
+        # Deploy IP address
+        public_ip = network.PublicIPAddress(
+            f"{self._name}_firewall_public_ip",
+            public_ip_address_name=f"{stack_name}-firewall-public-ip",
+            public_ip_allocation_method="Static",
+            resource_group_name=props.resource_group_name,
+            sku=network.PublicIPAddressSkuArgs(name="Standard"),
+            opts=opts,
+        )
+
+        # Get firewall subnet
+        snet_firewall = network.get_subnet_output(
+            subnet_name=props.subnet_firewall_name,
+            resource_group_name=props.resource_group_name,
+            virtual_network_name=props.subnet_firewall_virtual_network_name,
+        )
```

Comment:
This is saying, 'find me a subnet with this name, belonging to this virtual network, in this resource group'.

I assume this is a resource in the same stack?

We could pass the Python object about, but maybe this way keeps the code tidier when there are many resources?

Reply:
Yes. You can either create a `Subnet` as its own resource or as a subresource inside a `VirtualNetwork`. In order to stop the subnets being destroyed and recreated on a re-run we currently do the second one, which means there isn't a `Subnet` object to pass around. I've thought of a way we might be able to do the first one though - if it works, I'll update this logic.

Reply:
I see, I think I might have come across this before too.

If I remember correctly the vnet object has an attribute containing its subnets (although that might be Terraform).

Reply:
I do that [here](https://github.com/alan-turing-institute/bureau/blob/main/build/bureaubuild/infrastructure.py)

```python
    vnet = network.VirtualNetwork(
        'vnet',
        virtual_network_name='vnet',
        resource_group_name=resource_group.name,
        address_space=network.AddressSpaceArgs(
            address_prefixes=['10.0.0.0/16'],
        ),
        subnets=[network.SubnetArgs(
            name='default',
            address_prefix='10.0.0.0/24'
        )]
    )
```

and later

```python
        'focal': BuildVMArgs(
            resource_group=resource_group,
            image_reference=compute.ImageReferenceArgs(
                offer='0001-com-ubuntu-server-focal',
                publisher='canonical',
                sku='20_04-lts-gen2',
                version='latest',
            ),
            subnet=vnet.subnets[0],
            private_key=private_key
        ),
```

then in `BuildVM.init`

```python
        nic = network.NetworkInterface(
            f'{name}_nic',
            resource_group_name=args.resource_group.name,
            location=args.resource_group.location,
            ip_configurations=[
                network.NetworkInterfaceIPConfigurationArgs(
                    name='buildvmipconfig',
                    subnet=network.SubnetArgs(id=args.subnet.id),
                    public_ip_address=network.PublicIPAddressArgs(
                        id=public_ip.id
                    )
                )
            ],
            opts=child_opts,
        )
```

So it looks like `vnet.subnets` is a list of `pulumi_azure_native.network.subnet` rather than `SubnetArgs`.

Reply:
Yes, but they're ordered rather than named. I suppose we could match up the ordering the named return values?

Reply:
Oh, yes that does make it a bit tricker.

Could do something like

```python
def get_subnet(subnets: [subnet], name: str) -> Optional(subnet):
    for subnet in subnets:
        if subnet.name == name:
            return subnet
    return None
```

Reply:
Done by running `get_subnet` and adding the `GetSubnetResult` as an output from the component.